File: storage/db-core/src/main/java/com/coffee/gu/stamp/StampCouponUsageRepositoryImpl.java
package com.coffee.gu.stamp;

import com.querydsl.jpa.impl.JPAQueryFactory;
import org.springframework.stereotype.Repository;

import java.util.List;

import static com.coffee.gu.stamp.QStampCouponUsageEntity.*;

@Repository
public class StampCouponUsageRepositoryImpl implements StampCouponUsageRepository {

    private final StampCouponUsageJpaRepository stampCouponUsageJpaRepository;
    private final JPAQueryFactory jpaQueryFactory;

    public StampCouponUsageRepositoryImpl(StampCouponUsageJpaRepository stampCouponUsageJpaRepository, JPAQueryFactory jpaQueryFactory) {
        this.stampCouponUsageJpaRepository = stampCouponUsageJpaRepository;
        this.jpaQueryFactory = jpaQueryFactory;
    }

    @Override
    public StampCouponUsage save(StampCouponUsage stampCouponUsage) {
        return stampCouponUsageJpaRepository.save(StampCouponUsageEntity.from(stampCouponUsage)).toModel();
    }

    @Override
    public List<StampCouponUsage> saveAll(List<StampCouponUsage> stampCouponUsages) {
        return stampCouponUsageJpaRepository.saveAll(
                        stampCouponUsages.stream()
                                .map(StampCouponUsageEntity::from)
                                .toList()
                )
                .stream()
                .map(StampCouponUsageEntity::toModel)
                .toList();
    }

    @Override
    public List<StampCouponUsage> findAllByStampIdIn(List<Long> stampIds) {
        return jpaQueryFactory.selectFrom(stampCouponUsageEntity)
                .where(stampCouponUsageEntity.stampId.in(stampIds))
                .fetch()
                .stream()
                .map(StampCouponUsageEntity::toModel)
                .toList();
    }
}

File: storage/db-core/src/main/kotlin/com/coffee/gu/stamp/StampCouponUsageRepositoryImpl.kt
package com.coffee.gu.stamp

import com.coffee.gu.stamp.QStampCouponUsageEntity.stampCouponUsageEntity
import com.querydsl.jpa.impl.JPAQueryFactory
import org.springframework.stereotype.Repository

@Repository
class StampCouponUsageRepositoryImpl(
    private val stampCouponUsageJpaRepository: StampCouponUsageJpaRepository,
    private val jpaQueryFactory: JPAQueryFactory,
) : StampCouponUsageRepository {

    override fun save(stampCouponUsage: StampCouponUsage): StampCouponUsage {
        return stampCouponUsageJpaRepository.save(StampCouponUsageEntity.from(stampCouponUsage)).toModel()
    }

    override fun saveAll(stampCouponUsages: List<StampCouponUsage>): List<StampCouponUsage> {
        return stampCouponUsageJpaRepository.saveAll(
            stampCouponUsages.map { StampCouponUsageEntity.from(it) }
        ).map { it.toModel() }
    }

    override fun findAllByStampIdIn(stampIds: List<Long>): List<StampCouponUsage> {
        return jpaQueryFactory.selectFrom(stampCouponUsageEntity)
            .where(stampCouponUsageEntity.stampId.`in`(stampIds))
            .fetch()
            .map { it.toModel() }
    }
}
